File: src/main/java/xyz/chunkstories/api/net/Packet.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.net;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import xyz.chunkstories.api.exceptions.PacketProcessingException;

/** Packets are atomic data bits used to communicate over the network */
public abstract class Packet {
	public Packet() {

	}

	/** Called at send time, be consistent with the data you send, and give yourself
	 * a way to know how many bytes to expect if it's a variable length */
	public abstract void send(PacketDestinator destinator, DataOutputStream out, PacketSendingContext context) throws IOException;

	/** Called at reception, has to read the exact number of bytes sent or bad stuff
	 * happens ! */
	public abstract void process(PacketSender sender, DataInputStream in, PacketReceptionContext context) throws IOException, PacketProcessingException;
}

File: src/main/java/xyz/chunkstories/api/net/Packet.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.net

import xyz.chunkstories.api.Engine
import xyz.chunkstories.api.player.Player
import xyz.chunkstories.api.server.UserConnection
import xyz.chunkstories.api.world.GameInstance
import xyz.chunkstories.api.world.World
import java.io.DataInputStream
import java.io.DataOutputStream

typealias PacketId = Int

abstract class Packet(val engine: Engine) {
    abstract fun send(dos: DataOutputStream)
    abstract fun receive(dis: DataInputStream, user: UserConnection?)
}

abstract class PacketWorld(val world: World): Packet(world.gameInstance.engine) {
    final override fun receive(dis: DataInputStream, user: UserConnection?) = throw Exception("Call with the player argument instead")
    abstract fun receive(dis: DataInputStream, player: Player?)
}